impeller/entity/contents/filters/morphology_filter_contents.cc

```diff
@@ -0,0 +1,195 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/entity/contents/filters/morphology_filter_contents.h"
+
+#include <cmath>
+
+#include "impeller/entity/contents/content_context.h"
+#include "impeller/entity/contents/contents.h"
+#include "impeller/renderer/render_pass.h"
+#include "impeller/renderer/sampler_library.h"
+
+namespace impeller {
+
+DirectionalMorphologyFilterContents::DirectionalMorphologyFilterContents() =
+    default;
+
+DirectionalMorphologyFilterContents::~DirectionalMorphologyFilterContents() =
+    default;
+
+void DirectionalMorphologyFilterContents::SetRadius(Radius radius) {
+  radius_ = radius;
+}
+
+void DirectionalMorphologyFilterContents::SetDirection(Vector2 direction) {
+  direction_ = direction.Normalize();
+  if (direction_.IsZero()) {
+    direction_ = Vector2(0, 1);
+  }
+}
+
+void DirectionalMorphologyFilterContents::SetMorphType(MorphType morph_type) {
+  morph_type_ = morph_type;
+}
+
+std::optional<Snapshot> DirectionalMorphologyFilterContents::RenderFilter(
+    const FilterInput::Vector& inputs,
+    const ContentContext& renderer,
+    const Entity& entity,
+    const Rect& coverage) const {
+  using VS = MorphologyFilterPipeline::VertexShader;
+  using FS = MorphologyFilterPipeline::FragmentShader;
+
+  //----------------------------------------------------------------------------
+  /// Handle inputs.
+  ///
+
+  if (inputs.empty()) {
+    return std::nullopt;
+  }
+
+  auto input_snapshot = inputs[0]->GetSnapshot(renderer, entity);
+  if (!input_snapshot.has_value()) {
+    return std::nullopt;
+  }
+
+  if (radius_.radius < kEhCloseEnough) {
+    return input_snapshot.value();
+  }
+
+  auto transformed_radius = entity.GetTransformation().TransformDirection(
+      direction_ * radius_.radius);
+
+  auto transformed_radius_length = transformed_radius.GetLength();
+
+  // A matrix that rotates the snapshot space such that the direction is
+  // +X.
+  auto texture_rotate =
+      Matrix::MakeRotationZ(transformed_radius.Normalize().AngleTo({1, 0}));
```

Comment:
Flipping the texture to face +X direction was only needed in the blur filter for downsampling optimizations and this complexity can be avoided for this filter. I'd recommend taking a look at the blur filter's code before the recent downsampling optimization. The UV mapping situation was a bit simpler and `Snapshot` has a utility that can be used to compute the UVs here to great effect!

Reply:
Done.